### PR TITLE
Adds a utility to outer product and contraction of tensors as per  #509

### PR DESCRIPTION
**Pull request by bibek22**
Regarding #509 

@ritzvik Please have a look.

### CODE REVIEW

**Comment by bibek22** [comment]
I'm not sure which line is failing the check. Please suggest a better name for the function. `product` or `tensor_product` would probably be better. Also, is this the right file to place this function?

**Comment by shreyasbapat** [comment]
Why is it testing on Python 3.5 ?

**Comment by shreyasbapat** [comment]
Right. This needs a rebase :)

**Comment by bibek22** [comment]
> Right. This needs a rebase :)

Hey, I've pushed the changes. See if that does it.

File: src/einsteinpy/symbolic/tensor.py [hunk 1]
```diff
@@ -66,6 +66,51 @@ def chain_config_change():
     return chain_config_change()
 
 
+def multiply(t, s, i=None, j=None):
+    """product of tensors t and s (optionally) followed by summation of i and j
```
Comment (shreyasbapat) [inline src/einsteinpy/symbolic/tensor.py:70]: Docstring needs some serious changes. I will review thoroughly tomorrow Also, you will have to add tests to the PR

**Comment by bibek22** [comment]
I've made some serious changes and added a simple test case. Please have a look.

**Comment by bibek22** [comment]
It's the second to last item on the list. "Native function for Tensor Product"

File: src/einsteinpy/symbolic/tensor.py [hunk 2]
```diff
@@ -66,6 +66,54 @@ def chain_config_change():
     return chain_config_change()
 
 
+def tensor_product(t, s, i=None, j=None):
```
Comment (shreyasbapat) [inline src/einsteinpy/symbolic/tensor.py:69]:
```suggestion
def tensor_product(tensor1, tensor2, i=None, j=None):
```

File: src/einsteinpy/symbolic/tensor.py [hunk 3]
```diff
@@ -66,6 +66,54 @@ def chain_config_change():
...
+    """product of tensors ``t`` and ``s``
```
Comment (shreyasbapat) [inline src/einsteinpy/symbolic/tensor.py:70]: Docstring has to be corrected.

File: src/einsteinpy/symbolic/tensor.py [hunk 4]
```diff
@@ -66,6 +66,54 @@ def chain_config_change():
...
+
+    Parameters
+    ----------
+    t, s : ~einsteinpy.symbolic.BaseRelativityTensor
```
Comment (shreyasbapat) [inline src/einsteinpy/symbolic/tensor.py:74]: Parameters should be written separately. No shortcuts

File: src/einsteinpy/symbolic/tensor.py [hunk 5]
```diff
@@ -66,6 +66,54 @@ def chain_config_change():
...
+        Tensors product ``t`` and ``s``
+    i, j : ~int
+        contract  ``i``th index of ``t`` and ``j``th index of ``s``
```
Comment (shreyasbapat) [inline src/einsteinpy/symbolic/tensor.py:77]: Smilarly here.

File: src/einsteinpy/symbolic/tensor.py [hunk 6]
```diff
@@ -66,6 +66,54 @@ def chain_config_change():
...
 
 
+def tensor_product(t, s, i=None, j=None):
+    """product of tensors ``t`` and ``s``
+
+    Parameters
+    ----------
+    t, s : ~einsteinpy.symbolic.BaseRelativityTensor
+        Tensors product ``t`` and ``s``
+    i, j : ~int
```
Comment (shreyasbapat) [inline src/einsteinpy/symbolic/tensor.py:76]:
```suggestion
    i : int
        Brief description, optional
    j : int
```

**Comment by shreyasbapat** [comment]
You will need to write a better test.

**Comment by bibek22** [comment]
> You will need to write a better test.
alright. Could you suggest how I could improve it?

**Comment by shreyasbapat** [comment]
> > You will need to write a better test.
> 
> alright. Could you suggest how I could improve it?

You should open the CI Log! Once you find which line isn't tested, you'll have to write one for it. And then all checks will be :heavy_check_mark:

**Comment by bibek22** [comment]
> > > You will need to write a better test.
> > 
> > 
> > alright. Could you suggest how I could improve it?
> 
> You should open the CI Log! Once you find which line isn't tested, you'll have to write one for it. And then all checks will be heavy_check_mark

I'm sorry, I am not finding the log. Which one are you refering to? Is it on one of the ci/circleci ?

**Comment by shreyasbapat** [comment]
There are only two checks that have failed. Nvm, I pulled the link for you [here](https://codecov.io/gh/einsteinpy/einsteinpy/compare/d4f162f3c7b8ab5c87ac7fcdc1488a8f31b68252...787060bf59e366f51dbec9ae8ea32ec0b117d3b6/diff).

**Comment by shreyasbapat** [comment]
Great!

File: src/einsteinpy/symbolic/tensor.py [hunk 7]
```diff
@@ -66,6 +66,56 @@ def chain_config_change():
     return chain_config_change()
 
 
+def tensor_product(tensor1, tensor2, i=None, j=None):
+    """tensor product of ``tensor1`` and ``tensor2``
```
Comment (shreyasbapat) [inline src/einsteinpy/symbolic/tensor.py:70]:
```suggestion
    """
    Tensor Product of ``tensor1`` and ``tensor2``
```

File: src/einsteinpy/symbolic/tensor.py [hunk 8]
```diff
@@ -66,6 +66,56 @@ def chain_config_change():
...
+
+    Parameters
+    ----------
+    tensor1 : ~einsteinpy.symbolic.BaseRelativityTensor
+    tensor2 : ~einsteinpy.symbolic.BaseRelativityTensor
+    i : int
```
Comment (shreyasbapat) [inline src/einsteinpy/symbolic/tensor.py:76]:
```suggestion
    i : int, optional
```

File: src/einsteinpy/symbolic/tensor.py [hunk 9]
```diff
@@ -66,6 +66,56 @@ def chain_config_change():
...
+        contract ``i``th index of ``tensor1`` (optional)
```
Comment (shreyasbapat) [inline src/einsteinpy/symbolic/tensor.py:77]:
```suggestion
        contract ``i``th index of ``tensor1``
```

File: src/einsteinpy/symbolic/tensor.py [hunk 10]
```diff
@@ -66,6 +66,56 @@ def chain_config_change():
...
+    j : int
```
Comment (shreyasbapat) [inline src/einsteinpy/symbolic/tensor.py:78]:
```suggestion
    j : int, optional
```

File: src/einsteinpy/symbolic/tensor.py [hunk 11]
```diff
@@ -66,6 +66,56 @@ def chain_config_change():
...
+        contract ``j``th index of ``tensor2`` (optional)
```
Comment (shreyasbapat) [inline src/einsteinpy/symbolic/tensor.py:79]:
```suggestion
        contract ``j``th index of ``tensor2``
```

File: src/einsteinpy/tests/test_symbolic/test_tensor.py [hunk 12]
```diff
@@ -162,6 +180,16 @@ def test_ValueError2():
     obj = BaseRelativityTensor(test_list, [x, y, z])
 
 
+@xfail(raises=ValueError, strict=True)
+def test_ValueError3():
```
Comment (shreyasbapat) [inline src/einsteinpy/tests/test_symbolic/test_tensor.py:184]:
```suggestion
def test_tensorproduct_ValueError():
```

**Comment by shreyasbapat** [comment]
Rebase! You will face some merge conflicts!

File: src/einsteinpy/symbolic/tensor.py [hunk 13]
```diff
@@ -66,6 +66,56 @@ def chain_config_change():
...
+            raise ValueError(
+                "Index summation not allowed between %s and %s indices"
+                % (tensor1.config[i], tensor2.config[j])
+            )
+
+        product = simplify(tensorcontraction(product, (i, len(tensor1.config) + j)))
+
+        con = tensor1.config[:i] + tensor1.config[i + 1 :]
+        fig = tensor2.config[:j] + tensor2.config[j + 1 :]
+        newconfig = con + fig
```
Comment (ritwiksha) [inline src/einsteinpy/symbolic/tensor.py:107]: The variable naming made by day :laughing:

**Pull request merged**